### PR TITLE
[IMP] payment,sale: allow specifying the acquirer in a payment link

### PR DESCRIPTION
**Pull request by Demesmaeker**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/payment/wizards/payment_link_wizard.py [hunk 1]
```diff
@@ -41,7 +41,13 @@ def default_get(self, fields):
     link = fields.Char(string='Payment Link', compute='_compute_values')
     description = fields.Char('Payment Ref')
     access_token = fields.Char(compute='_compute_values')
-    company_id = fields.Many2one('res.company', compute='_compute_company')
+    company_id = fields.Many2one('res.company', store=True, compute='_compute_company')
+    payment_acquirer = fields.Many2one(
```
Comment (fmdl) [inline addons/payment/wizards/payment_link_wizard.py:45]: `payment_acquirer_id`

File: addons/payment/wizards/payment_link_wizard.py [hunk 2]
```diff
@@ -42,21 +40,18 @@ def default_get(self, fields):
     description = fields.Char('Payment Ref')
     access_token = fields.Char(compute='_compute_values')
     company_id = fields.Many2one('res.company', compute='_compute_company')
-    payment_acquirer_id = fields.Many2one(
-        string="Payment acquirer",
-        comodel_name="payment.acquirer",
-        index=True,
-        domain=[('state', 'in', ['enabled', 'test'])],
+    payment_acquirer = fields.Selection(
```
Comment (fmdl) [inline addons/payment/wizards/payment_link_wizard.py:43]: Why put a selection, if a company after lot of aquirer it is horrible to choose.

Many2one is better (you can search). If you want forgot edit or create in the view you can use :
```xml
<field name="payment_acquirer_id" options="{'no_create':True,'no_edit':True}"/>
```

Comment (chevalierv) [reply addons/payment/wizards/payment_link_wizard.py:43]: According to our backlog, the user shall have the choice to select "all" acquirers and a many2one doesn't allows to do that

File: addons/payment/wizards/payment_link_wizard.py [hunk 3]
```diff
@@ -7,8 +7,6 @@
 from odoo.tools import float_compare
 
 from odoo.addons.payment import utils as payment_utils
-
-
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:11]: Is this a mistake?

File: addons/payment/wizards/payment_link_wizard.py [hunk 4]
```diff
@@ -62,17 +63,60 @@ def _compute_values(self):
...
+            if len(self._get_payment_acquirer_available()) > 2:
+                link.has_multiple_acquirer = True
+            else:
+                link.has_multiple_acquirer = False
+
+    def _get_payment_acquirer_available(self):
+        """ Specify available acquirers in the selection field.
+
+        :return: a list of tuples, the first one being the 'all' one, and then
+        one tuple for each available acquirer, with state enabled or test.
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:81]: This is overly specific IMO. This could be shortened to something like `:return: The selection list of available acquirers`.

File: addons/payment/wizards/payment_link_wizard.py [hunk 5]
```diff
@@ -62,17 +63,60 @@ def _compute_values(self):
...
+            related_document = self.env[link.res_model].browse(link.res_id)
+            link.company_id = related_document.company_id
+
+    @api.depends('payment_acquirer')
+    def _compute_has_multiple_acquirer(self):
+        for link in self:
+            if len(self._get_payment_acquirer_available()) > 2:
+                link.has_multiple_acquirer = True
+            else:
+                link.has_multiple_acquirer = False
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:75]:
```suggestion
            link.has_multiple_acquirer = len(self._get_payment_acquirer_available()) > 2
```

File: addons/payment/wizards/payment_link_wizard.py [hunk 6]
```diff
@@ -62,17 +63,60 @@ def _compute_values(self):
...
+
+    def _get_payment_acquirer_available(self):
+        """ Specify available acquirers in the selection field.
+
+        :return: a list of tuples, the first one being the 'all' one, and then
+        one tuple for each available acquirer, with state enabled or test.
+        :rtype: list(tuple(id, name), ...)
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:82]:
```suggestion
        :rtype: list[tuple]
```

File: addons/payment/wizards/payment_link_wizard.py [hunk 7]
```diff
@@ -62,17 +63,60 @@ def _compute_values(self):
...
+        ) == -1:
+            raise ValidationError(_("Please set an amount smaller than %s.") % (self.amount_max))
+        if self.amount <= 0:
+            raise ValidationError(_("The value of the payment amount must be positive."))
 
     def _generate_link(self):
         for payment_link in self:
             related_document = self.env[payment_link.res_model].browse(payment_link.res_id)
-            base_url = related_document.get_base_url()  # Don't generate links for the wrong website
+            base_url = related_document.get_base_url() # Don't generate links for the wrong website
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:112]:
```suggestion
            base_url = related_document.get_base_url()  # Don't generate links for the wrong website
```
Keep the 2 space chars before an inline comment in Python.

File: addons/payment/wizards/payment_link_wizard.py [hunk 8]
```diff
@@ -62,17 +63,60 @@ def _compute_values(self):
...
+            payment_acquirer_url = f'&acquirer_id={payment_link.payment_acquirer}' \
+                if payment_link.payment_acquirer != 'all' else ''
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_link_wizard.py:114]: Couldn't this be part of `payment_link.link`'s computation? I feed that it adds more confusion than it tries to remove here.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Quick validation pass:
- Commit message: while the title is well written, the body should be used to explain the *why* rather than the *what*. The latter being described in full length by the diff.
In other words, "The dropdown allows the user to pre-select an available acquirer in the payment_link_wizard" can easily be figured out by reading the code. But why is that dropdown added?
Tip: The behavior behind it was already implemented and used by salesmen who had to add `acquirer_id=X` in the URL manually. Additionally, it wasn't very discoverable for the end-user.
- I see that you reorganized the methods in the wizards. Although it's a good idea to follow the guidelines, it's already a bit more delicate when the methods being moved are not impacted by this commit. It will add lots of friction to forward-ports for "nothing".
Instead, only move methods that you are also reworking to limit cases of forward-port conflicts to when a conflict was unavoidable anyway.

File: addons/payment/wizards/payment_link_wizard.py (hunk 8 above)
Comment (chevalierv) [reply addons/payment/wizards/payment_link_wizard.py:112]: I don't see why it would be better suit in `payment_link.link`'s computation because it would implied to create a new variable to be then accessed in `_generate_link`, for the same result because we would still need to check if the acquirer_id exists to put it in the url. I feel that compute the f-string in `payment_link.link`'s computation will bring more confusion since it will seems like the variable came from nowhere in `_generate_link`

Comment (AntoineVDV) [reply addons/payment/wizards/payment_link_wizard.py:112]: We have a misunderstanding. I'm suggesting to move `f'&acquirer_id={payment_link.payment_acquirer}' if payment_link.payment_acquirer != 'all' else ''` a few lines below, that's all. Like this:
```python
payment_link.link = f'{base_url}/payment/pay' \
    f'?reference={urls.url_quote(payment_link.description)}' \
    f'&amount={payment_link.amount}' \
    f'&currency_id={payment_link.currency_id.id}' \
    f'&partner_id={payment_link.partner_id.id}' \
    f'&company_id={payment_link.company_id.id}' \
    f'&acquirer_id={payment_link.payment_acquirer}' if payment_link.payment_acquirer != 'all' else '' \
    f'&access_token={payment_link.access_token}'
```
Or at least decouple the param name `acquirer_id` from its value. It's a bit weird to have all GET params listed below except for one that you must look up somewhere else. Also, "`payment_acquirer_url`" doesn't make much sense here as a variable name.

**Comment by AntoineVDV** [comment]
Unrequesting myself from review until the task is passing Testing BE.

File: addons/payment/wizards/payment_link_wizard.py [hunk 9]
```diff
@@ -42,15 +42,33 @@ def default_get(self, fields):
...
+        compute='_compute_has_multiple_acquirer',
+    )
 
     @api.onchange('amount', 'description')
     def _onchange_amount(self):
-        if float_compare(self.amount_max, self.amount, precision_rounding=self.currency_id.rounding or 0.01) == -1:
+        if float_compare(
+            self.amount_max, self.amount,
+            precision_rounding=self.currency_id.rounding or 0.01
+        ) == -1:
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:66]: Maybe I'm missing sthg, but I think this is an unrelated diff

File: addons/payment/wizards/payment_link_wizard.py [hunk 10]
```diff
@@ -42,15 +42,33 @@ def default_get(self, fields):
     description = fields.Char('Payment Ref')
     access_token = fields.Char(compute='_compute_values')
     company_id = fields.Many2one('res.company', compute='_compute_company')
+    available_acquirer_ids = fields.Many2many(
+        comodel_name='payment.acquirer',
+        string="Payment acquirers available",
+        compute='_compute_payment_acquiers',
+        compute_sudo=True,
+    )
+    payment_acquirer_id = fields.Many2one(
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:51]:
```suggestion
    acquirer_id = fields.Many2one(
```

Since the first part of model names refer to a module, we tend to use only the second part (I should add this to the guidelines :D)

File: addons/payment/wizards/payment_link_wizard.py [hunk 11]
```diff
@@ -42,15 +42,33 @@ def default_get(self, fields):
...
+        comodel_name='payment.acquirer',
+        string="Payment acquirer selected",
+        domain="[('id', 'in', available_acquirer_ids)]"
+    )
+    has_multiple_acquirer = fields.Boolean(
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:56]:
```suggestion
    has_multiple_acquirers = fields.Boolean(
```

File: addons/payment/wizards/payment_link_wizard.py [hunk 12]
```diff
@@ -62,8 +80,24 @@ def _compute_values(self):
     @api.depends('res_model', 'res_id')
     def _compute_company(self):
         for link in self:
-            record = self.env[link.res_model].browse(link.res_id)
-            link.company_id = record.company_id if 'company_id' in record else False
+            related_document = self.env[link.res_model].browse(link.res_id)
+            link.company_id = related_document.company_id
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:84]: Is this change still needed ?

File: addons/payment/wizards/payment_link_wizard.py [hunk 13]
```diff
@@ -62,8 +80,24 @@ def _compute_values(self):
...
+
+    @api.depends('company_id', 'partner_id', 'currency_id')
+    def _compute_payment_acquiers(self):
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:87]:
```suggestion
    def _compute_payment_acquirers(self):
```

File: addons/sale/wizard/sale_payment_link.py [hunk 14]
```diff
@@ -24,6 +24,16 @@ def default_get(self, fields):
...
         return res
 
+    @api.depends('company_id', 'partner_id', 'currency_id')
+    def _compute_payment_acquiers(self):
+        for link in self:
+            link.available_acquirer_ids = link.env['payment.acquirer']._get_compatible_acquirers(
+                company_id=link.company_id.id,
+                partner_id=link.partner_id.id,
+                currency_id=link.currency_id.id,
+                sale_order_id=link.res_id
```
Comment (Feyensv) [inline addons/sale/wizard/sale_payment_link.py:34]: link.res_id could be an account_move id :thinking:

File: addons/sale/wizard/sale_payment_link.py [hunk 15]
```diff
@@ -35,6 +45,7 @@ def _generate_link(self):
                                     f'?reference={urls.url_quote(payment_link.description)}' \
                                     f'&amount={payment_link.amount}' \
                                     f'&sale_order_id={payment_link.res_id}' \
+                                    f'{"&acquirer_id=" + str(payment_link.payment_acquirer_id.id) if payment_link.has_multiple_acquirer and payment_link.payment_acquirer_id.id else "" }' \
```
Comment (Feyensv) [inline addons/sale/wizard/sale_payment_link.py:48]: Is the check on `has_multiple_acquirer` useful here ?
The `acquirer_id` cannot be specified if there aren't multiple acquirers (and if it is, it's a custo and not really a problem to specify the acquirer)

**Review by Feyensv — COMMENTED** [review]
A lot of nitpicking/style comments (only one little mistake ^^).
Good job :+1:

File: addons/sale/wizard/sale_payment_link.py [hunk 16]
```diff
@@ -24,6 +24,19 @@ def default_get(self, fields):
...
+        for link in self:
+            kwargs = {
+                'company_id': self.company_id.id,
+                'partner_id': self.partner_id.id,
+                'currency_id': self.currency_id.id,
+            }
+            if link.res_model == 'sale.order':
+                kwargs['sale_order_id'] = link.res_id
+            link.available_acquirer_ids = link.env['payment.acquirer']._get_compatible_acquirers(
+                **kwargs)
```
Comment (Feyensv) [inline addons/sale/wizard/sale_payment_link.py:38]: Leave the management to super as much as possible ;)
(and do not use self in a for x in self loop)
```suggestion
        sale_links = self.filtered(lambda link: link.res_model == 'sale.order')
        super(PaymentLinkWizard, self-sale_links)._compute_payment_acquirers(self)
        for link in sale_links:
            link.available_acquirer_ids = link.env['payment.acquirer']._get_compatible_acquirers(
                company_id=link.company_id.id,
                partner_id=link.partner_id.id,
                currency_id=link.currency_id.id,
                sale_order_id=link.res_id)
```

File: addons/payment/wizards/payment_link_wizard_views.xml [hunk 17]
```diff
@@ -13,8 +13,11 @@
                         <field name="partner_id" invisible="1"/>
                         <field name="partner_email" invisible="1"/>
                         <field name="amount_max" invisible="1"/>
+                        <field name="available_acquirer_ids" invisible="1"/>
+                        <field name="has_multiple_acquirers" invisible="1"/>
                         <field name="description"/>
                         <field name="amount"/>
+                        <field name="acquirer_id" help="Let blank to select all acquirers" placeholder="Let blank to select all acquirers" options="{'no_open': True, 'no_create': True}" attrs="{'invisible':[('has_multiple_acquirers', '=', False)]}"/>
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard_views.xml:20]: Specify the help on the python field, not in the view ;)
And do not hesitate to indent xml lines when they are long (overly long lines are a nightmare in python but also in xml).

File: addons/payment/wizards/payment_link_wizard.py [hunk 18]
```diff
@@ -42,6 +42,21 @@ def default_get(self, fields):
     description = fields.Char('Payment Ref')
     access_token = fields.Char(compute='_compute_values')
     company_id = fields.Many2one('res.company', compute='_compute_company')
+    available_acquirer_ids = fields.Many2many(
+        comodel_name='payment.acquirer',
+        string="Payment Acquirers Available",
+        compute='_compute_payment_acquirers',
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:48]: Should have said it before, but the guidelines for compute methods are `_compute + field name = _compute_available_acquirer_ids`

```suggestion
        compute='_compute_available_acquirer_ids',
```

**Review by Feyensv — COMMENTED** [review]
Another batch of nitpicking/guidelines (+ the sale override wasn't updated, it won't be called if the method name isn't the same ^^)

File: addons/sale/wizard/sale_payment_link.py [hunk 19]
```diff
@@ -24,6 +24,17 @@ def default_get(self, fields):
             })
         return res
 
+    @api.depends('company_id', 'partner_id', 'currency_id')
+    def _compute_available_acquirer_ids(self):
+        sale_links = self.filtered(lambda link: link.res_model == 'sale.order')
+        super(PaymentLinkWizard, self-sale_links)._compute_available_acquirer_ids(self)
```
Comment (Feyensv) [inline addons/sale/wizard/sale_payment_link.py:30]: Mea culpa, the super call shouldn't receive an argument :man_facepalming: 
```suggestion
        super(PaymentLinkWizard, self-sale_links)._compute_available_acquirer_ids()
```

**Comment by Feyensv** [comment]
@robodoo r+
image

**Comment by Feyensv** [comment]
@robodoo r-
(Pierre wants "Leave blank" instead of "Let blank" :D)

**Comment by Feyensv** [comment]
@robodoo r+
:crossed_fingers:
